---
title: "RNA-Seq -- Differential Gene Expression Analysis"
author: "vaibhakv"
date: "2023-06-27"
output: 
  github_document:
    html_preview: false
---

## Experimental setup
* Saccharomyces cerevisiae was used as a model to study the mechanism of endogenous H2S that promoted the growth rate of yeast.
* The growth of fungi is controlled by several factors, one of which is signaling molecules, such as hydrogen sulfide (H2S), which was traditionally regarded as a toxic gas without physiological function. However, recent studies have revealed that H2S is produced enzymatically and endogenously in several species, where it serves as a gaseous signaling molecule performing a variety of critical biological functions. 
* However, the influence of this endogenous H2S on the biological activities occurring within the pathogenic fungi, such as transcriptomic and phenotypic alternations, has not been elucidated so far. Therefore, the present study was aimed to decipher this concern by utilizing S-propargyl-cysteine (SPRC) as a novel and stable donor of H2S and Saccharomyces cerevisiae as a fungal model.
* In this analysis, we'll find out the differentially expressed genes between two conditions with three replicates each -
  + 3 samples in the presence of SPRC -
    + SRR13978640
    + SRR13978641
    + SRR13978642
  + 3 samples in the absence of SPRC (control)
    + SRR13978643
    + SRR13978644
    + SRR13978645
    
```{r library_calls, message=FALSE}
library(tidyverse)
library(Rsubread)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(biomaRt)
library(ggrepel)
```

### featureCounts

Use featureCounts to generate the count table

```{r featureCounts, message=FALSE}
df <- featureCounts(files = c("data/output/SRR13978640.aln/SRR13978640Aligned.sortedByCoord.out.bam", 
                    "data/output/SRR13978641.aln/SRR13978641Aligned.sortedByCoord.out.bam",  
                    "data/output/SRR13978642.aln/SRR13978642Aligned.sortedByCoord.out.bam",
                    "data/output/SRR13978643.aln/SRR13978643Aligned.sortedByCoord.out.bam",
                    "data/output/SRR13978644.aln/SRR13978644Aligned.sortedByCoord.out.bam",
                    "data/output/SRR13978645.aln/SRR13978645Aligned.sortedByCoord.out.bam"),
                        annot.ext = "genome/Saccharomyces_cerevisiae.R64-1-1.109.gtf", 
                        isGTFAnnotationFile = TRUE,
                        isPairedEnd = TRUE,
                        nthreads = 4)

```

Save the count table into a variable and give the column names
```{r colnames}
raw_counts <- df$counts
colnames(raw_counts) <- paste0("SRR1397864",0:5) 
```

Create the metadata for samples
```{r metadata}
metadata <- tibble(sample = paste0("SRR1397864",0:5), 
                   condition = rep(c("SPRC", "control"), each=3))

```

### DESeqDataSetFromMatrix
Then we make a DESeq2 object 
```{r DESeq_object}
dds <- DESeqDataSetFromMatrix(countData = raw_counts, colData = metadata, design = ~ condition )
```

Visualizing the counts
```{r assay}
head(assay(dds)) 
```

We can filter out the genes with low counts
```{r filter}
keep <- rowSums(counts(dds)) > 10
dds <-  dds[keep,]
```

Estimating size factors and getting normalized counts
```{r size_factors}
sizeFactors <- estimateSizeFactors(dds)
normalized_counts <- counts(sizeFactors, normalized=TRUE)
```

We transform the data in order to minimize differences between samples for rows with small counts, and which helps in normalizing the data with respect to library size. 

* We can use 2 types of transformation
  + `vst`
  + `rlog`

#### Variance stabilizing transformation
```{r vst}
vst <-  varianceStabilizingTransformation(dds)
plotPCA(vst)
```

#### Regularized log transformation

```{r rlog}
rlog_dds <- rlog(dds)
plotPCA(rlog_dds)
```
As you can see `rlog` transformation is more robust than `vst`.

### Heatmap of samples

So for plotting the heatmap of sample distances, we transpose the `rlog_dds` matrix and calculate distances between them

Then we plot the heatmap.

```{r sample_heatmap}
sampleDist <- dist(t(assay(rlog_dds)))
sampleDistMatrix <- as.matrix(sampleDist)
rownames(sampleDistMatrix) <- dds$condition

pheatmap(sampleDistMatrix, 
         clustering_distance_rows = sampleDist,
         clustering_distance_cols = sampleDist,
         cols = colors)
```


### Running DESeq

We finally run `DESEq` on our original deseq object `dds` and not the transformed ones. Cause the DESeq does the transformation

```{r DESeq2_run}
res_dds <- DESeq(dds)
res <- results(res_dds)

```

We ran `DESeq` and stored the results into `res`

Here, you can see what the column names of `res` indicate. And also the `summary` of each column

##### Column names

```{r mcols_deseq}
mcols(res)
```

##### Summary

```{r deseq_summary}
summary(res)
```

#### Extracting significant genes 

In order to make a **heatmap** we'll have to set some **thresholds** first to extract only the top genes.
These include both *upregulated* and *downregulated*

```{r significant_genes}
res_sig <- subset(res, padj < 0.05 & baseMean > 100 & abs(log2FoldChange) > 1)
```

Most *upregulated*

```{r top_sig}
head(res_sig[order(res_sig$log2FoldChange),])
```

Most *downregulated*

```{r bottom_sig}
head(res_sig[order(res_sig$log2FoldChange, decreasing = TRUE),])
```

### MA plot

MA plots are commonly used to represent log fold-change versus mean expression between two treatments
They are scatter plots with **base-2 log fold-change** along the *y-axis* and **normalized mean expression** along the *x-axis*

First we have to **shrink** the log2fold changes before plotting. 

##### What is shrinkage?

In short, it looks at the largest fold changes that are not due to low counts and uses these to inform a prior distribution. So the large fold changes from genes with lots of statistical information are not shrunk, while the imprecise fold changes are shrunk.

This allows you to compare all estimated LFC across experiments. `DESeq2` does this by default. We're just doing it to make the MA plot

```{r MA_plot}
res_shrink <- lfcShrink(res_dds, coef = "condition_SPRC_vs_control", type = "apeglm")

plotMA(res_shrink, ylim = c(-3, 3))
```


```{r }
gene_ids_sig <- rownames(res_sig)
gene_ids <- rownames(res)
```

### bioMart
```{r messege=FALSE}
listEnsembl()
ensembl <- useEnsembl(biomart = "genes")
```

Selecting the **dataset**
```{r messege=FALSE}
listDatasets(ensembl) 
bio.mart <- useMart("ensembl", "scerevisiae_gene_ensembl")
```

Listing the **attributes** and **filters**
```{r messege=FALSE}
listAttributes(bio.mart)
listFilters(bio.mart) 
```

Getting the data from bioMart
```{r messege=FALSE}
bm_attr <- c("ensembl_gene_id", "external_gene_name", 
           "chromosome_name", "start_position", "end_position",
           "description")

bm_res <- getBM(attributes = bm_attr, 
      filters = c("ensembl_gene_id"), 
      values = gene_ids, mart = bio.mart)

bm_res_sig <- getBM(attributes = bm_attr, 
                filters = c("ensembl_gene_id"), 
                values = gene_ids_sig, mart = bio.mart)

bm_res_sig
```

#### Most significant Differentially Expressed Genes
```{r sig_gene_name}
bm_res_sig$external_gene_name
```

### Heatmap
Preparation for the Heatmap

```{r heatmap_matrix}
sig_matrix <- normalized_counts[rownames(res_sig),]
sig_matrix.z <- t(apply(sig_matrix, 1, scale))
rownames(sig_matrix.z) <- bm_res_sig$external_gene_name
colnames(sig_matrix.z) <- colnames(sig_matrix)
```

```{r heatcolors}
heat_colors <- colorRampPalette(brewer.pal(6, "PuOr"))(100)
```

```{r heatmap}
pheatmap(mat = sig_matrix.z, 
         cluster_rows = T, 
         cluster_cols = T, 
         labels_col = colnames(sig_matrix.z), 
         labels_row = rownames(sig_matrix.z),
         color = heat_colors,
         height = 40)
```

Preparation for the **Volcano plot**
```{r }
res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
```

Getting the **gene name** instead of ensembl ids
```{r }
res_tb <- res_tb %>% 
  mutate(gene_labels = bm_res$external_gene_name) 
```

Setting the **threshold**
```{r }
res_tb <- res_tb %>% 
  mutate(threshold = padj < 0.05 & abs(log2FoldChange) > 1) 
```

### Volcano plot
```{r volcano_plot}
ggplot(res_tb, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(colour = threshold)) +
  geom_text_repel(aes(label = gene_labels), alpha = 0.5) +
  geom_vline(xintercept = c(1, -1), linetype = "longdash", alpha = 0.3) +
  geom_hline(yintercept = 2, linetype = "longdash", alpha = 0.3) +
  scale_color_manual(values = c("#333366", "#CC9900")) +
  ggtitle("SPRC vs control - differentially expressed genes") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.2), hjust = 0.5),
        axis.title = element_text(size = rel(1.1)))
```

